From 54855569ceeb6c31fa50b2490e7a98f7fab305ce Mon Sep 17 00:00:00 2001
From: "Dr. Greg" <greg@enjellic.com>
Date: Sun, 7 Feb 2021 12:54:31 -0600
Subject: [PATCH 16/50] Inherit event measurement base from previous namespace.

With this patch a subordinate namespace inherits the event
measurement base from its parent namespace.

Previously the use of an event namespace map that was allocated
with kzalloc was used.  This caused a subordinate namespace to
revert back to using an all NULL array which violates the
principal of least surprise.
---
 security/integrity/ima/ima_events.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/security/integrity/ima/ima_events.c b/security/integrity/ima/ima_events.c
index e5d481529850..458abbd661f1 100644
--- a/security/integrity/ima/ima_events.c
+++ b/security/integrity/ima/ima_events.c
@@ -1934,6 +1934,13 @@ struct ima_events_namespace *ima_copy_events_ns(unsigned long flags,
 	}
 	memcpy(ns->map->AI_events, AI_events, sizeof(AI_events));
 
+	if (events_ns)
+		memcpy(ns->map->base, events_ns->map->base, WP256_DIGEST_SIZE);
+	else
+		memcpy(ns->map->base, init_ima_events_ns.map->base,
+		       WP256_DIGEST_SIZE);
+
+	/* Setup the update pseudo-file for this namespace. */
 	if (snprintf(bufr, sizeof(bufr), "%s%u", update, ns->ns.inum) >=
 	    sizeof(bufr)) {
 		retn = -ENOMEM;
-- 
2.31.1

